feat: Реализовано 2 слоя GateWay(с валидацией входных данных) и Serve…

HotSauceVL opened the pull request:
…r(с основной логикой приложения и взаимодействием с БД), добавлены настройки для docker

HotSauceVL committed:
feat: Реализовано 2 слоя GateWay(с валидацией входных данных) и Server(с основной логикой приложения и взаимодействием с БД), добавлены настройки для docker

HotSauceVL committed:
feat: Не успеваю, отправил, чтобы не идти в академ

HotSauceVL commented:
Извиняюсь заранее, что отправил, можно сразу назад на доработку) Если есть возможность подсказать, почему фейлятся тесты на гите, было бы супер)

avfyodorov reviewed (commented):
Добрый день, Владимир!

Да, сразу на доработку. Тесты. Извините что задержал, просто сегодня какой-то сумасшедший день. Сегодня прошла как раз информация, что они там с тестами что-то меняют... перестраивают.  Конкретики, как всегда, нет. Вам лучше было бы напрямую поговорить с наставником, Он должен быть в курсе.

HotSauceVL committed:
refactor: изменил dto в gateway

HotSauceVL committed:
refactor: попытка убрать ошибку IllegalState Failed to load ApplicationContex

HotSauceVL committed:
refactor: попытка убрать ошибку IllegalState Failed to load ApplicationContex

HotSauceVL committed:
style: убрал неиспользуемый импорт

HotSauceVL committed:
style: убрал неиспользуемые импорты

HotSauceVL committed:
refactor: исправил ошибки работы запроса ItemRequestServiceImpl.getAllItemRequestsCreatedAnotherUsers

HotSauceVL commented:
Ура) спасибо) все ожило) Ваша подсказка о новых тестах, помогла догадаться нормально допилить pom.xml, а там и остальные недоработки пофиксил) пребывая в состоянии рука лицо, почему я раньше до этого не дошел) отправляю рабочий проект) спасибо)

avfyodorov commented on gateway/pom.xml:
```diff
@@ -0,0 +1,48 @@
+        </dependency>
+
+        <dependency>
+            <groupId>org.hibernate.validator</groupId>
```
Этa зависимость нужна?

avfyodorov commented on gateway/src/main/java/ru/practicum/shareit/exception/GatewayExceptionHandler.java:
```diff
@@ -0,0 +1,31 @@
+import javax.validation.ConstraintViolationException;
+
+@RestControllerAdvice
+public class GatewayExceptionHandler {
```
Сюда хорошо бы добавить логирование. 
И ещё добавить хендлер для обработки всех необработанных исключений, ловить будем Throwable, а возвращать статус 500 INTERNAL_SERVER_ERROR
Так же, для полноты, можно добавить обработчик для исключения, которое выбрасывается при валидации данных с помощью аннотаций - MethodArgumentNotValidException, возвращать будет статус 400 BAD_REQUEST, так как ошибка связана с некорректными данными от пользователя).

avfyodorov commented on gateway/src/main/java/ru/practicum/shareit/item/GateWayItemController.java:
```diff
@@ -0,0 +1,82 @@
+                                         @Positive @RequestParam(name = "size", required = false,
+                                                 defaultValue = "20") int size) {
+        log.info("Search items by text {}, user {}, from {}, size {}", text, userId, from, size);
+        return itemClient.searchByText(text, userId, from, size);
```
Лучше было бы прямо здесь проверить входной параметр text. И, если текст запроса пустой, или недопустимый, например, состоит из пробелов, то прямо отсюда и вернуть результат, не обращаться к серверу.

avfyodorov commented on server/pom.xml:
```diff
@@ -0,0 +1,70 @@
+		<dependency>
+			<groupId>org.postgresql</groupId>
+			<artifactId>postgresql</artifactId>
+			<version>42.5.0</version>
```
Зависимости, для которых требуется указание версий следует располагать в блоке <dependencyManagement> в родительском POM, а в дочерних указывать их уже без ее указания

avfyodorov reviewed (commented):
Добрый день, Владимир!

Работа хорошая, конечно, есть несколько замечаний, но они мелкие.

HotSauceVL committed:
refactor: исправлены ошибки ревью, исправлены зависимости в pom.xml, добавлена валидация text в методе GateWayItemController.searchByText
feat: добавлено исклбчение ServerErrorException, реализован хендлер для ServerErrorException и MethodArgumentNotValidException

HotSauceVL committed:
refactor: исправлена валидация text в методе GateWayItemController.searchByText, согласно тестам при пустом text возвращается 200 и пустой список

HotSauceVL committed:
style: удалены лишние импорты

HotSauceVL committed:
style: удалены лишние пробелы

avfyodorov reviewed (commented):
Добрый день, Владимир!

Замечаний нет, работа принята.
 Теперь диплом, успехов!
